### PR TITLE
cdc: Add description about protocol behavior change.

### PR DESCRIPTION
**Pull request by 3AceShowHand**
### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [x] v8.0 (TiDB 8.0 versions)
- [x] v7.6 (TiDB 7.6 versions)
- [x] v7.5 (TiDB 7.5 versions)
- [x] v7.4 (TiDB 7.4 versions)
- [x] v7.1 (TiDB 7.1 versions)
- [x] v6.5 (TiDB 6.5 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: ticdc/ticdc-sink-to-cloud-storage.md [hunk 1]
```diff
@@ -344,3 +344,7 @@ schema 文件中对 Enum/Set 类型的定义如下：
     "ColumnType":"{ENUM/SET}",
 }
 ```
+
+### 行为变更说明
```
Comment (Oreoxmt) [inline ticdc/ticdc-sink-to-cloud-storage.md:348]:
```suggestion
## 行为变更说明
```

File: ticdc/ticdc-sink-to-kafka.md [hunk 2]
```diff
@@ -511,3 +511,7 @@ Kafka 消费者会收到一条含有大消息在外部存储服务中的地址
 ```
 
 `key` 和 `value` 分别存有编码后的大消息，该消息原本应该发送到 Kafka 消息中的对应字段。消费者可以通过解析这两部分的数据，还原大消息的内容。
+
+### 行为变更说明
```
Comment (Oreoxmt) [inline ticdc/ticdc-sink-to-kafka.md:515]:
```suggestion
## 行为变更说明
```

File: ticdc/ticdc-sink-to-cloud-storage.md [hunk 3]
```diff
@@ -344,3 +344,7 @@ schema 文件中对 Enum/Set 类型的定义如下：
     "ColumnType":"{ENUM/SET}",
 }
 ```
+
+## 行为变更说明
+
+从 v6.5.3, v7.1.1 开始。如果 Update 事件的主键或者非空唯一键的列值发生改变，该条事件将会被拆分为 Delete 和 Insert 两条事件。这样做的目的是因为 CSV 协议在输出 Update 的事件内容时，不输出旧值，通过该方法可以解决该问题。[config(ticdc): enable-old-value always false if using avro or csv as the encoding protocol](https://github.com/pingcap/tiflow/pull/9079)
```
[the remark this answers is not in the document]
Comment (3AceShowHand) [reply ticdc/ticdc-sink-to-cloud-storage.md]: 这里把 update 拆了之后，old value 不就以 delete message 的形式被输出了么。

File: ticdc/ticdc-sink-to-kafka.md [hunk 4]
```diff
@@ -511,3 +511,7 @@ Kafka 消费者会收到一条含有大消息在外部存储服务中的地址
 ```
 
 `key` 和 `value` 分别存有编码后的大消息，该消息原本应该发送到 Kafka 消息中的对应字段。消费者可以通过解析这两部分的数据，还原大消息的内容。
+
+## 行为变更说明
+
+从 v6.5.3, v7.1.1, v7.2.0-DMR 开始。如果 Update 事件的主键或者非空唯一键的列值发生改变，该条事件将会被拆分为 Delete 和 Insert 两条事件。这样做的目的是为了避免因为下游消费者组对多个 Kafka Topic Partitions 消费进度不同，导致的数据不一致问题。[config(ticdc): enable-old-value always false if using avro or csv as the encoding protocol](https://github.com/pingcap/tiflow/pull/9079)
```
Comment (Oreoxmt) [inline ticdc/ticdc-sink-to-kafka.md:517]:
```suggestion
从 v6.5.3、v7.1.1 和 v7.2.0 开始，TiCDC 在处理 Update 事件时，如果事件的主键或者非空唯一索引的列值发生改变，则会将该条事件拆分为 Delete 和 Insert 两条事件。这是为了避免当下游消费者组处理多个 Kafka Topic partition 时，由于消费进度不同导致数据不一致的问题。详情见 GitHub issue [#9086](https://github.com/pingcap/tiflow/issues/9086)。
```

File: ticdc/ticdc-sink-to-cloud-storage.md [hunk 5]
```diff
@@ -344,3 +344,7 @@ schema 文件中对 Enum/Set 类型的定义如下：
     "ColumnType":"{ENUM/SET}",
 }
 ```
+
+## 行为变更说明
+
+从 v6.5.3, v7.1.1, v7.2.0-DMR 开始。如果 Update 事件的主键或者非空唯一键的列值发生改变，该条事件将会被拆分为 Delete 和 Insert 两条事件。这样做的目的是因为 CSV 协议在输出 Update 的事件内容时，不输出旧值，通过该方法可以解决该问题。[config(ticdc): enable-old-value always false if using avro or csv as the encoding protocol](https://github.com/pingcap/tiflow/pull/9079)
```
[the remark this answers is not in the document]
Comment (Oreoxmt) [reply ticdc/ticdc-sink-to-cloud-storage.md:350]: 建议使用 issue 链接

File: ticdc/ticdc-behavior-change.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+insert into t values (1, 1);
+
+update t set a = 2 where a = 1;
+```
+
+上述例子中，主键 a = 1 被修改为 a = 2。如果不做拆分。使用 CSV 和 AVRO 协议时，只能看到新值 a = 2，无法知道旧值 a = 1。下游消费者可能只插入了新值 2，而没有删除旧值 1。使用 Index value dispatcher 时，插入记录 (1, 1) 的事件可能被发送到 Partition 0，而变更事件 (2, 1) 可能被发送到 Partition 1，如果 Partition 1 的消费进度快于 Partition 0，则可能因为在下游数据系统中找不到相应数据而出错。因此，TiCDC 会将该条事件拆分为 Delete 和 Insert 两条事件，即删除记录 (1, 1)，写入记录 (2, 1)，前者被写入到 Partition 0，后者写入到 Partition 1，无论消费者进度如何，都能被消费成功。
+
+## 含有多条 Update 变更的事物拆分
+
+从 v6.5.4、v7.1.2 和 v7.4.0 开始，使用 MySQL Sink 时。对于一个含有多条变更的事物，如果 Update 事件的主键或者非空唯一索引的列值发生改变，TiCDC 会将该其拆分为 Delete 和 Insert 两条事件，并且将所有事件排序，保证 Delete 事件在 Insert 事件之前。详情见 GitHub issue[#9430](https://github.com/pingcap/tiflow/pull/9437)
```
Comment (CharlesCheung96) [inline ticdc/ticdc-behavior-change.md:31]: 这个变更对所有 sink 类型都生效

Comment (3AceShowHand) [reply ticdc/ticdc-behavior-change.md:31]: 从源代码上来看，主要是针对 mysql sink 而言。

对于其他类型的 sink，在处理事物这个环节，已经没有 key update 事件了，当然也有排序。

确实是对所有 sink 类型都有效，我添加一下。

File: ticdc/ticdc-behavior-change.md [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+```sql
+create table t (a int primary key, b int);
+
+insert into t values (1, 1);
+
+update t set a = 2 where a = 1;
+```
+
+上述例子中，主键 a = 1 被修改为 a = 2。如果不做拆分。使用 CSV 和 AVRO 协议时，只能看到新值 a = 2，无法知道旧值 a = 1。下游消费者可能只插入了新值 2，而没有删除旧值 1。使用 Index value dispatcher 时，插入记录 (1, 1) 的事件可能被发送到 Partition 0，而变更事件 (2, 1) 可能被发送到 Partition 1，如果 Partition 1 的消费进度快于 Partition 0，则可能因为在下游数据系统中找不到相应数据而出错。因此，TiCDC 会将该条事件拆分为 Delete 和 Insert 两条事件，即删除记录 (1, 1)，写入记录 (2, 1)，前者被写入到 Partition 0，后者写入到 Partition 1，无论消费者进度如何，都能被消费成功。
```
Comment (CharlesCheung96) [inline ticdc/ticdc-behavior-change.md:27]:
```suggestion
上述例子中，主键 a = 1 被修改为 a = 2。如果不做拆分：
*  使用 CSV 和 AVRO 协议时，只能看到新值 a = 2，无法知道旧值 a = 1。下游消费者可能只插入了新值 2，而没有删除旧值 1。
* 使用 Index value dispatcher 时，插入记录 (1, 1) 的事件可能被发送到 Partition 0，而变更事件 (2, 1) 可能被发送到 Partition 1，如果 Partition 1 的消费进度快于 Partition 0，则可能因为在下游数据系统中找不到相应数据而出错。因此，TiCDC 会将该条事件拆分为 Delete 和 Insert 两条事件，其中，删除记录 (1, 1) 被发送到 Partition 0，写入记录 (2, 1) 被发送到 Partition 1，无论消费者进度如何，都能被消费成功。
```

File: ticdc/ticdc-behavior-change.md [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+## 含有多条 Update 变更的事物拆分
+
+从 v6.5.4、v7.1.2 和 v7.4.0 开始。对于一个含有多条变更的事物，如果 Update 事件的主键或者非空唯一索引的列值发生改变，TiCDC 会将该其拆分为 Delete 和 Insert 两条事件，并且将所有事件排序，保证 Delete 事件在 Insert 事件之前。详情见 GitHub issue[#9430](https://github.com/pingcap/tiflow/pull/9437)
+
+该变更主要为了解决如下问题：
+
+* 使用 MySQL Sink 直接写入这两条事件到下游，下会出现主键冲突的问题，导致 changefeed 报错。
```
Comment (CharlesCheung96) [inline ticdc/ticdc-behavior-change.md:35]:
```suggestion
* 使用 MySQL Sink 直接写入这两条事件到下游，会出现主键冲突的问题，导致 changefeed 报错。
```

**Review by CharlesCheung96 — APPROVED** [review]

File: releases/release-6.5.3.md [hunk 9]
```diff
@@ -11,6 +11,14 @@ TiDB 版本：6.5.3
 
 试用链接：[快速体验](https://docs.pingcap.com/zh/tidb/v6.5/quick-start-with-tidb) | [生产部署](https://docs.pingcap.com/zh/tidb/v6.5/production-deployment-using-tiup) | [下载离线包](https://cn.pingcap.com/product-community/?version=v6.5.3#version-list)
 
+## 兼容性变更
+
+### 行为变更
+
+* TiCDC
+
+    - 主键或者非空索引键发生改变的 Update 事件，拆分为 Delete 和 Insert 两条事件。[含有单条 Update 变更的事物拆分](/ticdc/ticdc-behavior-change.md#含有单条-update-变更的事物拆分)
```
Comment (Oreoxmt) [inline releases/release-6.5.3.md:20]:
```suggestion
- TiCDC 在处理 Update 事件时，如果事件的主键或者非空唯一索引的列值发生改变，则会将该条事件拆分为 Delete 和 Insert 两条事件。更多信息，请参考[用户文档](/ticdc/ticdc-behavior-change.md#含有单条-update-变更的事物拆分)。
```

File: releases/release-6.5.4.md [hunk 10]
```diff
@@ -16,6 +16,12 @@ TiDB 版本：6.5.4
 - 为了修复当使用 `Cursor Fetch` 获取大结果集时 TiDB 占用大量内存的问题，TiDB 会自动将结果集写入磁盘以释放内存资源 [#43233](https://github.com/pingcap/tidb/issues/43233) @[YangKeao](https://github.com/YangKeao)
 - 默认关闭 RocksDB 的周期性 compaction，使 TiKV RocksDB 的默认行为和 v6.5.0 之前的版本保持一致，避免在升级之后集中产生大量 compaction 影响系统的性能。同时，TiKV 新增 [`rocksdb.[defaultcf|writecf|lockcf].periodic-compaction-seconds`](https://docs.pingcap.com/zh/tidb/v6.5/tikv-configuration-file#periodic-compaction-seconds-从-v654-版本开始引入) 和 [`rocksdb.[defaultcf|writecf|lockcf].ttl`](https://docs.pingcap.com/zh/tidb/v6.5/tikv-configuration-file#ttl-从-v654-版本开始引入) 配置项，支持手动配置 RocksDB 的周期性 compaction [#15355](https://github.com/tikv/tikv/issues/15355) @[LykxSassinator](https://github.com/LykxSassinator)
 
+### 行为变更
+
+* TiCDC
+
+    - 对于一个含有多条变更的事物，如果 Update 事件的主键或者非空唯一索引的列值发生改变，TiCDC 会将该其拆分为 Delete 和 Insert 两条事件，并且将所有事件排序，保证 Delete 事件在 Insert 事件之前。[含有多条 Update 变更的事物拆分](/ticdc/ticdc-behavior-change.md#含有多条-update-变更的事物拆分)
```
Comment (Oreoxmt) [inline releases/release-6.5.4.md:23]:
```suggestion
- 对于包含多条变更的事务，如果 Update 事件的主键或者非空唯一索引的列值发生改变，TiCDC 会将该其拆分为 Delete 和 Insert 两条事件，并确保将所有事件有序，以保证 Delete 事件在 Insert 事件之前。更多信息，请参考[用户文档](/ticdc/ticdc-behavior-change.md#含有多条-update-变更的事物拆分)。
```

File: releases/release-7.1.1.md [hunk 11]
```diff
@@ -15,6 +15,12 @@ TiDB 版本：7.1.1
 
 - TiDB 新增 `tidb_lock_unchanged_keys` 变量用于控制是否对于没有修改的 key 加锁 [#44714](https://github.com/pingcap/tidb/issues/44714) @[ekexium](https://github.com/ekexium)
 
+### 行为变更
+
+* TiCDC
+
+    - 主键或者非空索引键发生改变的 Update 事件，拆分为 Delete 和 Insert 两条事件。[含有单条 Update 变更的事物拆分](/ticdc/ticdc-behavior-change.md#含有单条-update-变更的事物拆分)
```
Comment (Oreoxmt) [inline releases/release-7.1.1.md:22]:
```suggestion
- TiCDC 在处理 Update 事件时，如果事件的主键或者非空唯一索引的列值发生改变，则会将该条事件拆分为 Delete 和 Insert 两条事件。更多信息，请参考[用户文档](/ticdc/ticdc-behavior-change.md#含有单条-update-变更的事物拆分)。
```

File: releases/release-7.1.2.md [hunk 12]
```diff
@@ -20,6 +20,12 @@ TiDB 版本：7.1.2
 - 新增 TiCDC 配置项 [`sink.csv.binary-encoding-method`](/ticdc/ticdc-changefeed-config.md#ticdc-changefeed-配置文件说明)，控制 CSV 协议中二进制类型数据的编码方式，默认值为 `'base64'` [#9373](https://github.com/pingcap/tiflow/issues/9373) @[CharlesCheung96](https://github.com/CharlesCheung96)
 - 新增 TiCDC 配置项 [`large-message-handle-option`](/ticdc/ticdc-sink-to-kafka.md#处理超过-kafka-topic-限制的消息)。默认为空，即消息大小超过 Kafka Topic 的限制后，同步任务失败。设置为 "handle-key-only" 时，如果消息超过大小，只发送 handle key 以减少消息的大小；如果依旧超过大小，则同步任务失败 [#9680](https://github.com/pingcap/tiflow/issues/9680) @[3AceShowHand](https://github.com/3AceShowHand)
 
+### 行为变更
+
+* TiCDC
+
+    - 对于一个含有多条变更的事物，如果 Update 事件的主键或者非空唯一索引的列值发生改变，TiCDC 会将该其拆分为 Delete 和 Insert 两条事件，并且将所有事件排序，保证 Delete 事件在 Insert 事件之前。[含有多条 Update 变更的事物拆分](/ticdc/ticdc-behavior-change.md#含有多条-update-变更的事物拆分)
```
Comment (Oreoxmt) [inline releases/release-7.1.2.md:27]:
```suggestion
- 对于包含多条变更的事务，如果 Update 事件的主键或者非空唯一索引的列值发生改变，TiCDC 会将该其拆分为 Delete 和 Insert 两条事件，并确保将所有事件有序，以保证 Delete 事件在 Insert 事件之前。更多信息，请参考[用户文档](/ticdc/ticdc-behavior-change.md#含有多条-update-变更的事物拆分)。
```

File: releases/release-7.2.0.md [hunk 13]
```diff
@@ -170,6 +170,12 @@ TiDB 版本：7.2.0
 >
 > 以下为从 v7.1.0 升级至当前版本 (v7.2.0) 所需兼容性变更信息。如果从 v7.0.0 或之前版本升级到当前版本，可能也需要考虑和查看中间版本 release notes 中提到的兼容性变更信息。
 
+### 行为变更
+
+* TiCDC
+
+    - 主键或者非空索引键发生改变的 Update 事件，拆分为 Delete 和 Insert 两条事件。[含有单条 Update 变更的事物拆分](/ticdc/ticdc-behavior-change.md#含有单条-update-变更的事物拆分)
```
Comment (Oreoxmt) [inline releases/release-7.2.0.md:177]:
```suggestion
- TiCDC 在处理 Update 事件时，如果事件的主键或者非空唯一索引的列值发生改变，则会将该条事件拆分为 Delete 和 Insert 两条事件。更多信息，请参考[用户文档](/ticdc/ticdc-behavior-change.md#含有单条-update-变更的事物拆分)。
```

File: releases/release-7.4.0.md [hunk 14]
```diff
@@ -286,6 +286,10 @@ TiDB 版本：7.4.0
 
 - 新增表 [`information_schema.CHECK_CONSTRAINTS`](/information-schema/information-schema-check-constraints.md)，提高与 MySQL 8.0 的兼容性。
 
+* TiCDC
+
+    - 对于一个含有多条变更的事物，如果 Update 事件的主键或者非空唯一索引的列值发生改变，TiCDC 会将该其拆分为 Delete 和 Insert 两条事件，并且将所有事件排序，保证 Delete 事件在 Insert 事件之前。[含有多条 Update 变更的事物拆分](/ticdc/ticdc-behavior-change.md#含有多条-update-变更的事物拆分)
```
Comment (Oreoxmt) [inline releases/release-7.4.0.md:291]:
```suggestion
- 对于包含多条变更的事务，如果 Update 事件的主键或者非空唯一索引的列值发生改变，TiCDC 会将该其拆分为 Delete 和 Insert 两条事件，并确保将所有事件有序，以保证 Delete 事件在 Insert 事件之前。更多信息，请参考[用户文档](/ticdc/ticdc-behavior-change.md#含有多条-update-变更的事物拆分)。
```

File: ticdc/ticdc-behavior-change.md
Comment (Oreoxmt) [inline ticdc/ticdc-behavior-change.md:1]: 新增文件需要加入 TOC.md

File: ticdc/ticdc-behavior-change.md [hunk 15]
```diff
@@ -0,0 +1,54 @@
+---
+title: 行为变更说明
+summary: 记录 TiCDC changefeed 行为变更，说明变更原因，以及影响范围。
+---
```
Comment (Oreoxmt) [inline ticdc/ticdc-behavior-change.md:4]:
```suggestion
---
title: TiCDC 行为变更说明
summary: 介绍 TiCDC changefeed 的行为变更，说明变更原因以及影响范围。
---
```

File: ticdc/ticdc-behavior-change.md [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+
+# Update 事件拆分为 Delete 和 Insert 事件
+
+## 含有单条 Update 变更的事物拆分
```
Comment (Oreoxmt) [inline ticdc/ticdc-behavior-change.md:8]:
```suggestion
# TiCDC 行为变更说明

## 将 Update 事件拆分为 Delete 和 Insert 事件

### 含有单条 Update 变更的事务拆分
```
title 需要与一级标题保持一致

File: ticdc/ticdc-behavior-change.md [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+
+从 v6.5.3、v7.1.1 和 v7.2.0 开始，使用非 MySQL Sink 时。对于只含有一条 Update 变更的事物，如果 Update 事件的主键或者非空唯一索引的列值发生改变，TiCDC 会将该条事件拆分为 Delete 和 Insert 两条事件。详情见 GitHub issue [#9086](https://github.com/pingcap/tiflow/issues/9086)。
```
Comment (Oreoxmt) [inline ticdc/ticdc-behavior-change.md:10]:
```suggestion
从 v6.5.3、v7.1.1 和 v7.2.0 开始，使用非 MySQL Sink 时，对于仅包含一条 Update 变更的事务，如果 Update 事件的主键或者非空唯一索引的列值发生改变，TiCDC 会将该条事件拆分为 Delete 和 Insert 两条事件。详情见 GitHub issue [#9086](https://github.com/pingcap/tiflow/issues/9086)。
```

File: ticdc/ticdc-behavior-change.md [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+
+该变更主要为了解决如下问题：
+
+* CSV 和 AVRO 协议，只输出新值，不输出旧值。当主键或者非空唯一索引的列值发生改变时，消费者只能收到变化后的新值，不知道旧值，也就无法处理变更之前的值（如删除旧值）。
+* 使用 Index value dispatcher，将数据按照 Key 分发到不同的 Kafka partitions。下游消费者组中的多个消费者进程独立地消费Kafka Topic partition，由于消费进度不同，可能导致数据不一致的问题。
```
Comment (Oreoxmt) [inline ticdc/ticdc-behavior-change.md:15]:
```suggestion
* 在使用 CSV 和 AVRO 协议时，仅输出新值而不输出旧值。因此，当主键或者非空唯一索引的列值发生改变时，消费者只能接收到变化后的新值，无法得到旧值，导致无法处理变更前的值（例如删除旧值）。
* 在使用 Index value dispatcher 将数据按照 Key 分发到不同的 Kafka partition 时，下游的消费者组内多个消费者进程独立消费 Kafka Topic partition，由于消费进度不同，可能导致数据不一致的问题。
```

File: ticdc/ticdc-behavior-change.md [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+
+考虑如下 SQL：
+
+```sql
+create table t (a int primary key, b int);
+
+insert into t values (1, 1);
+
+update t set a = 2 where a = 1;
+```
```
Comment (Oreoxmt) [inline ticdc/ticdc-behavior-change.md:25]:
````suggestion
以如下 SQL 为例：

```sql
CREATE TABLE t (a INT PRIMARY KEY, b INT);
INSERT INTO t VALUES (1, 1);
UPDATE t SET a = 2 WHERE a = 1;
```
````

File: ticdc/ticdc-behavior-change.md [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+
+上述例子中，主键 a = 1 被修改为 a = 2。如果不做拆分：
```
Comment (Oreoxmt) [inline ticdc/ticdc-behavior-change.md:27]:
```suggestion
在上述示例中，主键 `a` 的值从 `1` 修改为 `2`。如果不将该 Update 事件进行拆分：
```

File: ticdc/ticdc-behavior-change.md [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+
+*  使用 CSV 和 AVRO 协议时，只能看到新值 a = 2，无法知道旧值 a = 1。下游消费者可能只插入了新值 2，而没有删除旧值 1。
+* 使用 Index value dispatcher 时，插入记录 (1, 1) 的事件可能被发送到 Partition 0，而变更事件 (2, 1) 可能被发送到 Partition 1，如果 Partition 1 的消费进度快于 Partition 0，则可能因为在下游数据系统中找不到相应数据而出错。因此，TiCDC 会将该条事件拆分为 Delete 和 Insert 两条事件，其中，删除记录 (1, 1) 被发送到 Partition 0，写入记录 (2, 1) 被发送到 Partition 1，无论消费者进度如何，都能被消费成功。
```
Comment (Oreoxmt) [inline ticdc/ticdc-behavior-change.md:30]:
```suggestion
* 在使用 CSV 和 AVRO 协议时，消费者仅能看到新值 `a = 2`，而无法得到旧值 `a = 1`。这可能导致下游消费者只插入了新值 `2`，而没有删除旧值 `1`。
* 在使用 Index value dispatcher 时，插入记录 `(1, 1)` 的事件可能被发送到 Partition 0，而变更事件 `(2, 1)` 可能被发送到 Partition 1。如果 Partition 1 的消费进度快于 Partition 0，则可能由于下游数据系统中找不到相应数据而导致出错。因此，TiCDC 会将该条 Update 事件拆分为 Delete 和 Insert 两条事件，其中，删除记录 `(1, 1)` 被发送到 Partition 0，写入记录 `(2, 1)` 被发送到 Partition 1，以确保无论消费者的进度如何，事件都能被消费成功。
```

File: ticdc/ticdc-behavior-change.md [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+
+## 含有多条 Update 变更的事物拆分
```
Comment (Oreoxmt) [inline ticdc/ticdc-behavior-change.md:32]:
```suggestion
### 含有多条 Update 变更的事务拆分
```

File: ticdc/ticdc-behavior-change.md [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+
+从 v6.5.4、v7.1.2 和 v7.4.0 开始。对于一个含有多条变更的事物，如果 Update 事件的主键或者非空唯一索引的列值发生改变，TiCDC 会将该其拆分为 Delete 和 Insert 两条事件，并且将所有事件排序，保证 Delete 事件在 Insert 事件之前。详情见 GitHub issue[#9430](https://github.com/pingcap/tiflow/pull/9437)
```
Comment (Oreoxmt) [inline ticdc/ticdc-behavior-change.md:34]:
```suggestion
从 v6.5.4、v7.1.2 和 v7.4.0 开始，对于一个含有多条变更的事务，如果 Update 事件的主键或者非空唯一索引的列值发生改变，TiCDC 会将该其拆分为 Delete 和 Insert 两条事件，并确保所有事件按照 Delete 事件在 Insert 事件之前的顺序进行排序。详情见 GitHub issue[#9430](https://github.com/pingcap/tiflow/pull/9437)。
```

File: ticdc/ticdc-behavior-change.md [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+
+该变更主要为了解决如下问题：
+
+* 使用 MySQL Sink 直接写入这两条事件到下游，会出现主键冲突的问题，导致 changefeed 报错。
```
Comment (Oreoxmt) [inline ticdc/ticdc-behavior-change.md:38]:
```suggestion
该变更主要为了解决当使用 MySQL Sink 直接将这两条事件写入下游时，可能会出现主键冲突的问题，从而导致 changefeed 报错。
```

File: ticdc/ticdc-behavior-change.md [hunk 25]
```diff
@@ -0,0 +1,54 @@
...
+create table t (a int primary key, b int);
+insert into t values (1, 1);
+insert into t values (2, 2);
+
+begin;
+update t set a = 1 where a = 3;
+update t set a = 2 where a = 1;
+update t set a = 3 where a = 2;
+commit;
+```
```
Comment (Oreoxmt) [inline ticdc/ticdc-behavior-change.md:52]:
````suggestion
以如下 SQL 为例：

```sql
CREATE TABLE t (a INT PRIMARY KEY, b INT);
INSERT INTO t VALUES (1, 1);
INSERT INTO t VALUES (2, 2);

BEGIN;
UPDATE t SET a = 1 WHERE a = 3;
UPDATE t SET a = 2 WHERE a = 1;
UPDATE t SET a = 3 WHERE a = 2;
COMMIT;
```
````

File: ticdc/ticdc-behavior-change.md [hunk 26]
```diff
@@ -0,0 +1,54 @@
...
+
+上述例子中给出的交换两行数据的主键的事物，执行了 3 条 SQL 语句，但是 TiCDC 只会收到 2 条 Update 变更事件，即将主键 1 变更为 2，将主键 2 变更为 1，这会导致 changefeed 报错。因此，TiCDC 会将这两条事件拆分为 4 条事件，即删除记录 (1, 1), (2, 2), 写入记录 (2, 1), (1, 2)。
```
Comment (Oreoxmt) [inline ticdc/ticdc-behavior-change.md:54]:
```suggestion
在上述示例中，通过执行三条 SQL 语句对两行数据的主键进行交换，但 TiCDC 只会接收到两条 Update 变更事件，即将主键 `a` 从 `1` 变更为 `2`，将主键 `a` 从 `2` 变更为 `1`，这可能会导致 changefeed 报错。

因此，TiCDC 会将这两条事件拆分为四条事件，即删除记录 `(1, 1)` 和 `(2, 2)` 以及写入记录 `(2, 1)` 和 `(1, 2)`。
```

**Review by flowbehappy — APPROVED** [review]

File: ticdc/ticdc-behavior-change.md [hunk 27]
```diff
@@ -0,0 +1,54 @@
...
+在上述示例中，主键 `a` 的值从 `1` 修改为 `2`。如果不将该 Update 事件进行拆分：
+
+* 在使用 CSV 和 AVRO 协议时，消费者仅能看到新值 `a = 2`，而无法得到旧值 `a = 1`。这可能导致下游消费者只插入了新值 `2`，而没有删除旧值 `1`。
+* 在使用 Index value dispatcher 时，插入记录 `(1, 1)` 的事件可能被发送到 Partition 0，而变更事件 `(2, 1)` 可能被发送到 Partition 1。如果 Partition 1 的消费进度快于 Partition 0，则可能由于下游数据系统中找不到相应数据而导致出错。因此，TiCDC 会将该条 Update 事件拆分为 Delete 和 Insert 两条事件，其中，删除记录 `(1, 1)` 被发送到 Partition 0，写入记录 `(2, 1)` 被发送到 Partition 1，以确保无论消费者的进度如何，事件都能被消费成功。
+
+### 含有多条 Update 变更的事务拆分
+
+从 v6.5.4、v7.1.2 和 v7.4.0 开始，对于一个含有多条变更的事务，如果 Update 事件的主键或者非空唯一索引的列值发生改变，TiCDC 会将该其拆分为 Delete 和 Insert 两条事件，并确保所有事件按照 Delete 事件在 Insert 事件之前的顺序进行排序。详情见 GitHub issue[#9430](https://github.com/pingcap/tiflow/pull/9437)。
+
+该变更主要为了解决当使用 MySQL Sink 直接将这两条事件写入下游时，可能会出现主键冲突的问题，从而导致 changefeed 报错。
```
Comment (hfxsd) [inline ticdc/ticdc-behavior-change.md:36]:
```suggestion
该变更主要为了解决当使用 MySQL Sink 直接将 Delete 和 Insert 这两条事件写入下游时，可能会出现主键冲突的问题，从而导致 changefeed 报错的问题。
```

File: ticdc/ticdc-behavior-change.md [hunk 28]
```diff
@@ -0,0 +1,54 @@
...
+```
+
+在上述示例中，主键 `a` 的值从 `1` 修改为 `2`。如果不将该 Update 事件进行拆分：
+
+* 在使用 CSV 和 AVRO 协议时，消费者仅能看到新值 `a = 2`，而无法得到旧值 `a = 1`。这可能导致下游消费者只插入了新值 `2`，而没有删除旧值 `1`。
+* 在使用 Index value dispatcher 时，插入记录 `(1, 1)` 的事件可能被发送到 Partition 0，而变更事件 `(2, 1)` 可能被发送到 Partition 1。如果 Partition 1 的消费进度快于 Partition 0，则可能由于下游数据系统中找不到相应数据而导致出错。因此，TiCDC 会将该条 Update 事件拆分为 Delete 和 Insert 两条事件，其中，删除记录 `(1, 1)` 被发送到 Partition 0，写入记录 `(2, 1)` 被发送到 Partition 1，以确保无论消费者的进度如何，事件都能被消费成功。
+
+### 含有多条 Update 变更的事务拆分
+
+从 v6.5.4、v7.1.2 和 v7.4.0 开始，对于一个含有多条变更的事务，如果 Update 事件的主键或者非空唯一索引的列值发生改变，TiCDC 会将该其拆分为 Delete 和 Insert 两条事件，并确保所有事件按照 Delete 事件在 Insert 事件之前的顺序进行排序。详情见 GitHub issue[#9430](https://github.com/pingcap/tiflow/pull/9437)。
```
Comment (Oreoxmt) [inline ticdc/ticdc-behavior-change.md:34]:
```suggestion
从 v6.5.4、v7.1.2 和 v7.4.0 开始，对于一个含有多条变更的事务，如果 Update 事件的主键或者非空唯一索引的列值发生改变，TiCDC 会将该其拆分为 Delete 和 Insert 两条事件，并确保所有事件按照 Delete 事件在 Insert 事件之前的顺序进行排序。详情见 GitHub issue [#9430](https://github.com/pingcap/tiflow/pull/9437)。
```

Comment (hfxsd) [inline ticdc/ticdc-behavior-change.md:34]:
```suggestion
从 v6.5.4、v7.1.2 和 v7.4.0 开始，对于一个含有多条变更的事务，如果 Update 事件的主键或者非空唯一索引的列值发生改变，TiCDC 会将该其拆分为 Delete 和 Insert 两条事件，并确保所有事件按照 Delete 事件在 Insert 事件之前的顺序进行排序。详情见 GitHub issue [#9430](https://github.com/pingcap/tiflow/issues/9430)。
```

File: ticdc/ticdc-behavior-change.md [hunk 29]
```diff
@@ -0,0 +1,54 @@
...
+INSERT INTO t VALUES (2, 2);
+
+BEGIN;
+UPDATE t SET a = 1 WHERE a = 3;
+UPDATE t SET a = 2 WHERE a = 1;
+UPDATE t SET a = 3 WHERE a = 2;
+COMMIT;
+```
+
+在上述示例中，通过执行三条 SQL 语句对两行数据的主键进行交换，但 TiCDC 只会接收到两条 Update 变更事件，即将主键 `a` 从 `1` 变更为 `2`，将主键 `a` 从 `2` 变更为 `1`，这可能会导致 changefeed 报错。
```
Comment (hfxsd) [inline ticdc/ticdc-behavior-change.md:52]: 示例中的 3 ，这里没有提。请确认是否要修改。
img_v3_028b_d58cabec-38e1-413c-aae2-72fa15f8b0dg

Comment (3AceShowHand) [reply ticdc/ticdc-behavior-change.md:52]: 不需要提，它只是一个中间语句，不会被下游用户感知到。

**Review by hfxsd — APPROVED** [review]

**Pull request merged**